### PR TITLE
CARRY: Add RHOAI manifests

### PR DESCRIPTION
**Pull request by z103cb**
**What this PR does / why we need it**:
Added the manifests to allow for deployment into a RHOAI.

**Which issue(s) this PR fixes** 
Closes: https://issues.redhat.com/browse/RHOAIENG-4787
**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/training/) included if any changes are user facing

### CODE REVIEW

File: manifests/rhoai/binding_admin_roles.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+kind: ClusterRoleBinding
```
Comment (z103cb) [inline manifests/rhoai/binding_admin_roles.yaml:1]: @astefanutti Is this needed ?

File: manifests/rhoai/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+# Adds namespace to all resources.
+namespace: opendatahub
+
+# Value of this field is prepended to the
+# names of all resources, e.g. a deployment named
+# "wordpress" becomes "alices-wordpress".
+# Note that it should also match with the prefix (text before '-') of the namespace
+# field above.
+namePrefix: kfto-
```
Comment (z103cb) [inline manifests/rhoai/kustomization.yaml:11]: @astefanutti, @tedhtchang and @jbusche Does this prefix make sense ?

File: manifests/rhoai/params.env [hunk 3]
```diff
@@ -0,0 +1 @@
+odh-training-operator-controller-image=docker.io/kubeflow/training-operator:v1-855e096
```
Comment (z103cb) [inline manifests/rhoai/params.env:1]: @astefanutti is there an quay image for this ?

File: manifests/rhoai/kubeflow-training-roles.yaml [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  name: kubeflow-training-admin
+  labels:
+#    rbac.authorization.kubeflow.org/aggregate-to-kubeflow-admin: "true"
+    rbac.authorization.k8s.io/aggregate-to-admin: "true"
+aggregationRule:
+  clusterRoleSelectors:
+    - matchLabels:
+#        rbac.authorization.kubeflow.org/aggregate-to-kubeflow-training-admin: "true"
+        rbac.authorization.k8s.io/aggregate-to-admin: "true"
+rules: []
```
Comment (zdtsw) [inline manifests/rhoai/kubeflow-training-roles.yaml:15]: should this be defined with which resource are needed?

File: manifests/rhoai/monitor.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Prometheus Pod Monitor (Metrics)
+apiVersion: monitoring.coreos.com/v1
+kind: PodMonitor
+metadata:
+  name: controller-manager-metrics-monitor
+  namespace: opendatahub
```
Comment (zdtsw) [inline manifests/rhoai/monitor.yaml:6]: this probably not needed

File: manifests/rhoai/params.env (hunk 3 above)
Comment (astefanutti) [reply manifests/rhoai/params.env:1]: Not yet. We can keep the upstream one in the short term until we have automated build for the container image.

File: manifests/rhoai/kubeflow-training-roles.yaml [hunk 6]
```diff
@@ -0,0 +1,107 @@
+#This file has been copied from  ../overlays/kubeflow
+#The original labels have ben commented out for documentation purposes
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
```
Comment (astefanutti) [inline manifests/rhoai/kubeflow-training-roles.yaml:4]: I don't think that ClusterRole is needed.

File: manifests/rhoai/binding_admin_roles.yaml (hunk 1 above)
Comment (astefanutti) [reply manifests/rhoai/binding_admin_roles.yaml:1]: I don't think that's needed.

File: manifests/rhoai/kustomization.yaml (hunk 2 above)
Comment (astefanutti) [reply manifests/rhoai/kustomization.yaml:11]: Based on the names from `base` I'd suggest `kubeflow-` as a prefix.

File: manifests/rhoai/monitor.yaml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# Prometheus Pod Monitor (Metrics)
+apiVersion: monitoring.coreos.com/v1
+kind: PodMonitor
```
Comment (astefanutti) [inline manifests/rhoai/monitor.yaml:3]: Ideally, we would remove the Service created for the metrics endpoint from base. Kueue `rhoai` overlay does that for reference.

File: manifests/rhoai/kubeflow-training-roles.yaml [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    - matchLabels:
+#        rbac.authorization.kubeflow.org/aggregate-to-kubeflow-training-admin: "true"
+        rbac.authorization.k8s.io/aggregate-to-admin: "true"
+rules: []
+
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: kubeflow-training-edit
```
Comment (astefanutti) [inline manifests/rhoai/kubeflow-training-roles.yaml:21]: If `kubeflow-` is the prefix, that should be `training-edit`.

**Comment by jbusche** [comment]
OK - I built a OC 4.14.17 cluster with ODH 2.9 and simple dsc and tried installing kfto, not quite sure what I'm exactly testing...
```
git clone https://github.com/z103cb/training-operator.git -b rhoai-manifests
cd training-operator/manifests/rhoai
kustomize build | oc apply -f -
```
Note, my IBM cluster always has trouble with pull rate limits from [docker.io](http://docker.io/), so I've put up the image into quay/jbusche and then configured it with this command:
```
oc set image deployment kfto-training-operator training-operator=quay.io/jbusche/training-operator:v1-855e096 -n opendatahub
```
I also had to trick the initcontainer like this to get around the rate limits too:
```
oc patch deployment kfto-training-operator -n opendatahub --type='json' -p='[{"op": "replace", "path": "/spec/template/spec/containers/0/command", "value": ["/manager",  "--pytorch-init-container-image=quay.io/jbusche/alpine:3.10"]}]'
```
Now I've deployed the simple.yaml to my own namespace demo-dsp and it's worked:
```
oc get pods,pytorchjobs -n demo-dsp
NAME                          READY   STATUS      RESTARTS   AGE
pod/demo-wb-0                 2/2     Running     0          16m
pod/pytorch-simple-master-0   0/1     Completed   0          4m6s
pod/pytorch-simple-worker-0   0/1     Completed   0          6m16s

NAME                                     STATE       AGE
pytorchjob.kubeflow.org/pytorch-simple   Succeeded   6m16s
```

File: manifests/rhoai/binding_admin_roles.yaml (hunk 1 above)
Comment (z103cb) [reply manifests/rhoai/binding_admin_roles.yaml:1]: Ok, will remove

File: manifests/rhoai/kustomization.yaml (hunk 2 above)
Comment (z103cb) [reply manifests/rhoai/kustomization.yaml:11]: Ok. will fix.

File: manifests/rhoai/params.env (hunk 3 above)
Comment (z103cb) [reply manifests/rhoai/params.env:1]: Ok.

File: manifests/rhoai/monitor.yaml (hunk 5 above)
Comment (z103cb) [reply manifests/rhoai/monitor.yaml:6]: Ok. will remove.

File: manifests/rhoai/monitor.yaml (hunk 7 above)
Comment (z103cb) [reply manifests/rhoai/monitor.yaml:3]: Ok, i will add patch to delete the service from base.

File: manifests/rhoai/kubeflow-training-roles.yaml (hunk 6 above)
Comment (z103cb) [reply manifests/rhoai/kubeflow-training-roles.yaml:4]: Ok, removing it for now.

**Comment by z103cb** [comment]
@jbusche if you can retest this now, greatly appreciated !

**Comment by jbusche** [comment]
LGTM @z103cb, I deployed on a new cluster and it worked.  I noticed the prefix has changed to kubeflow so I needed to adjust my image tweaks like this:
```
oc set image deployment kubeflow-training-operator training-operator=quay.io/jbusche/training-operator:v1-855e096 -n opendatahub

oc patch deployment kubeflow-training-operator -n opendatahub --type='json' -p='[{"op": "replace", "path": "/spec/template/spec/containers/0/command", "value": ["/manager",  "--pytorch-init-container-image=quay.io/jbusche/alpine:3.10"]}]'
```
and then my kfto operator came up.
```
oc get pods -n opendatahub |grep kubeflow
kubeflow-training-operator-65db576c9d-qmbt2        1/1     Running   0          14m
```
The deployment of a pytorchjob in a non-default namespace looked good:
```
oc get pytorchjobs,pods                                                     api.jim414rh.cp.fyre.ibm.com: Fri Mar 29 10:52:24 2024

NAME                                     STATE       AGE
pytorchjob.kubeflow.org/pytorch-simple   Succeeded   9m53s

NAME                          READY   STATUS	  RESTARTS   AGE
pod/demo-wb-0                 2/2     Running     0          49m
pod/pytorch-simple-master-0   0/1     Completed   0          9m53s
pod/pytorch-simple-worker-0   0/1     Completed   0          9m53s
```

File: manifests/rhoai/kubeflow-training-roles.yaml [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+      - list
+      - watch
+  - apiGroups:
+      - ""
+    resources:
+      - events
+    verbs:
+      - get
+      - list
+      - watch
```
Comment (astefanutti) [inline manifests/rhoai/kubeflow-training-roles.yaml:58]: I'd removed those rules.

**Review by astefanutti — APPROVED** [review]

**Pull request merged**